### PR TITLE
Add bootrom compiler flow

### LINKED ISSUE #227: Bootroms cannot be easily modified
This is a pretty minor issue, as the bootroms do work and probably won't need to change in future, but it would be nice if the raw files here worked to generate a fresh binary..

The input files in BootROMs/ can't be used to generate a fresh bootrom, so we are stuck with the bootrom as it is.
I can compile logo_compress.c, but using the program on the existing SameBoyLogo.png gives an assertion error.

A workaround is to use the SameBoy workflow to compile new bootroms if need be, but this currently does not readily support using different GBC logos (so we would be stuck with "SAMEBOY.")

### PR DESCRIPTION
**Pull request by maij**
This PR adds a workflow to create new, open-source DMG, SGB and CGB bootroms which are built into the core.
The reason I have worked on this is to resolve #227 and partially #192.

#192 was primarily resolved by PR #229, which allows you to boot into a CGB game even with a different palette selected (this palette is discarded). However, this functionality is not representative of the actual CGB bootrom, which prevents palette selection for all CGB games. Here, I have modified the boot rom to prevent the palette from changing when a CGB cart is loaded. This includes both CGB-only carts and DMG compatible CGB carts.

There was also some dead-code which allowed CGB games to run in DMG mode (on a CGB). This has now been removed as it is not implemented in the core.

As an added bonus, I modified the "GAME BOY" logo to be closer to what is in the original CGB bootrom.
Overall, I think it's worth being able to rebuild everything from scratch. Feedback is welcome, this may be an unexpected contribution.

Here is a build with the updated bootrom: [Gameboy_bootrom_20230116.zip](https://github.com/MiSTer-devel/Gameboy_MiSTer/files/10439908/Gameboy_bootrom_20230116.zip)

Todo:
- [ ] Test compilation on a Windows system
- [x] Remove all remaining Sameboy items
- [x] Squash commits
- [x] Add reference to SameBoy source license (source code is only slightly modified here.)
- [x] Add Readme to ./BootRoms
- [x] Update repo readme

### CODE REVIEW

**Comment by Kitrinx** [comment]
One thing to please watch for if you are implementing a GBC boot rom is the switch for GBA detection for a handful of games. It's not a key feature but it would be nice to have that with a third party rom.

Just to clarify one thing though, this BIOS is licensed for public use? I see nintendo trademarked logos included which could be a hazard for the project, so they should probably be replaced with "MiSTer" or something along those lines.

**Comment by maij** [comment]
There is an AGB switch in the cgb_boot.asm (set in agb_boot.asm) which creates a modified bootrom. I have compiled all of these and put them `./bin` if you would like to try it. We could in theory add a fourth bootrom to the `.mif` file, which would take another 2 kiB of memory in the core. You would then just add another OSD menu option to run in `AGB` mode.

The sameboy source is given with an MIT license, so we can use it so long as it is attributed correctly. I think the only place the Nintendo logo exists is in the cart roms themselves. The bootrom reads this off and renders it on screen. Is that what you mean? If so, we could remove the comments that say Nintendo, just to be safe.

**Comment by paulb-nl** [comment]
Games detect GBA mode by looking at bit 0 of register b which is set on a GBA and not set on a GBC. We don't need to add another bootrom just for that. In the past the core used the read value of $FF50 to enable fast boot in the bootrom and we could also do that to enable GBA mode. 

The new default DMG bootrom that you added uses the Sameboy logo animation. I don't know what people will think about that, since this is not the original animation. The Sameboy animation seems more appropriate as a separate dmg_boot_fast.bin file.

**Comment by paulb-nl** [comment]
I don't think there is source of the current DMG bootrom. It seems to be a modified original bootrom. The animation will have to be reimplemented.

Your code to read from $FF50 looks good but it should check for a specific value since $FF50 normally reads as $FF. The previous fast boot method used value $42.

File: BootROMs/src/cgb_boot.asm [hunk 1]
```diff
@@ -198,15 +191,23 @@ ELSE
     call PlaySound
 ENDC
     call Preboot
-
-; Will be filled with NOPs
+    ld a, [rBANK] ; Special MiSTer read register to give menu boot options
+    bit 0, a
+    jr z, .cgbmode
```
Comment (paulb-nl) [inline BootROMs/src/cgb_boot.asm:196]:
```suggestion
    cp $42
    jr nz, .cgbmode
```

File: Gameboy.sv [hunk 2]
```diff
@@ -357,7 +357,7 @@ hps_io #(.CONF_STR(CONF_STR), .WIDE(1)) hps_io
 
 	.buttons(buttons),
 	.status(status),
-	.status_menumask({sys_megaduck,using_real_cgb_bios,sgb_border_en,isGBC,cart_ready,sav_supported,|tint,gg_available}),
+	.status_menumask({4'b0,cart_ready,sav_supported,|tint,gg_available}),
```
Comment (paulb-nl) [inline Gameboy.sv:360]: Why did you remove `sys_megaduck`, `sgb_border_en` and `isGBC`? Those are in use.

File: Gameboy.sv [hunk 3]
```diff
@@ -541,11 +541,8 @@ cart_top cart (
 );
 
 reg [127:0] palette = 128'h828214517356305A5F1A3B4900000000;
-reg using_real_cgb_bios = 0;
 
 always @(posedge clk_sys) begin
-	if (cgb_boot_download)
-		using_real_cgb_bios <= 1;
```
Comment (paulb-nl) [inline Gameboy.sv:548]: You should not remove `using_real_cgb_bios`. That will remove the ability to use GBA mode with the original bootrom.

File: rtl/gb.v [hunk 4]
```diff
@@ -174,6 +174,10 @@ wire sel_FF73  = isGBC && cpu_addr == 16'hff73;            // unused register, a
 wire sel_FF74  = isGBC && isGBC_mode && (cpu_addr == 16'hff74); // unused register, all bits read/write, only in CGB mode
 wire sel_FF75  = isGBC && cpu_addr == 16'hff75;            // unused register, bits 4-6 read/write
 
+// Special MiSTer register for signalling to cpu bootrom
+// Normally this register is write-only so no
+wire sel_FF50  = isGBC && boot_rom_enabled && cpu_addr == 16'hff50;
```
Comment (paulb-nl) [inline rtl/gb.v:179]:
```suggestion
// Special MiSTer register for signaling GBA mode to cpu bootrom
// Normally this register is write-only so no
wire sel_FF50  = isGBC && boot_gba_en && boot_rom_enabled && cpu_addr == 16'hff50;
```

File: Gameboy.sv [hunk 5]
```diff
@@ -609,7 +606,7 @@ gb gb (
 
 	.nCS         ( nCS        ),
 
-	.boot_gba_en    ( status[37] && using_real_cgb_bios ),
+	.boot_gba_en    ( status[37] ),
```
Comment (paulb-nl) [inline Gameboy.sv:609]:
```suggestion
	.boot_gba_en    ( status[37] ),
	.real_cgb_boot  ( using_real_cgb_bios ),
```

File: rtl/gb.v [hunk 6]
```diff
@@ -928,38 +933,17 @@ wire [10:0] boot_wr_addr =
...
-		12'h0F9: boot_do_gba = 8'hE0;
-		12'h0FA: boot_do_gba = 8'h70;
-		12'h0FB: boot_do_gba = 8'h04;
-		12'h409: boot_do_gba = 8'h80;
-		12'h40A: boot_do_gba = 8'hFF;
-		default: boot_do_gba = boot_q;
-	endcase
-end
-
-assign boot_do = (isGBC & boot_gba_en) ? boot_do_gba : boot_q;
```
Comment (paulb-nl) [inline rtl/gb.v:962]: This should be restored but it should be `(isGBC & boot_gba_en & real_cgb_boot)`

File: BootROMs/src/cgb_boot.asm (hunk 1 above)
Comment (maij) [reply BootROMs/src/cgb_boot.asm:196]: I did read through your suggestion from your earlier comment, I was going to write a comment later explaining these changes and a few others (I am still working on a couple of things right now). Wasn't expecting a review to come in!

I had a think about this, and so long as the $FF50 register is overriden while `boot_rom_enabled` is active, we have complete control over the read/write accesses. cgb_boot.asm does not read from $FF50, except here. So really, we can use a single bit rather than an 8 bit literal. I don't see the benefit of comparing to $42. 

Does the real cgb bootrom read from $FF50, and is that a cause for concern?

File: Gameboy.sv (hunk 2 above)
Comment (maij) [reply Gameboy.sv:360]: I removed them from the OSD menu mask because those bits weren't being used. I can't see any option in the CORE_CONF_STRING that is masked by these bits.

File: Gameboy.sv (hunk 3 above)
Comment (maij) [reply Gameboy.sv:548]: This was a considered decision, that we can reverse if need be. 
I figured that it doesn't make much sense to inject code into the real CGB bios rom, when the new CGB rom built here can natively emulate AGB mode. It makes sense to me to reduce the complexity when an alternative solution is practical and easily available.

File: Gameboy.sv (hunk 2 above)
Comment (maij) [reply Gameboy.sv:360]: Ah, I got confused I only looked at the 'hide' options in the conf string but not the 'disable' options. My mistake.

File: BootROMs/src/cgb_boot.asm (hunk 1 above)
Comment (paulb-nl) [reply BootROMs/src/cgb_boot.asm:196]: Yes you are right. It doesn't really matter since the boot roms don't normally read it anyway.

File: Gameboy.sv (hunk 3 above)
Comment (paulb-nl) [reply Gameboy.sv:548]: Some people like to use the original CGB bootrom and in that case GBA mode should also work so we need to keep both methods.

I actually want to add more complexity because I realized that there is no check that it is indeed the original bootrom that is loaded and not `cgb_boot_fast.bin` for example. So I would like to verify either by checking a few bytes or summing all bytes and verifying the sum.

**Comment by maij** [comment]
Latest build:
[Gameboy_bootrom_20230130.zip](https://github.com/MiSTer-devel/Gameboy_MiSTer/files/10540919/Gameboy_bootrom_20230130.zip)

**Comment by paulb-nl** [comment]
It looks good to me. Are you still planning to squash commits?

**Comment by maij** [comment]
Just squashed commits, kept the main ones. The only thing I can't do is test the make on Windows.

**Comment by maij** [comment]
I am happy for this to be merged if others are happy. If someone wants to work on the Windows side of things and the Makefile isn't 100% correct, they can modify it.

I have made a final addition to the boot which switches the final blue colour when AGB mode is selected, same as what Sameboy does. I have made a final test build and everything is working as intended.

[Gameboy_20230211.zip](https://github.com/MiSTer-devel/Gameboy_MiSTer/files/10713105/Gameboy_20230211.zip)
